R/zchunk_LB134.Diet_Rfao.R

```diff
@@ -0,0 +1,384 @@
+      select(-value_fhy) %>%
+      bind_rows(filter(L134.pcFood_kcald_R_Dmnd_Y, year < max(HISTORICAL_YEARS)))->
+      L134.pcFood_kcald_R_Dmnd_Y
+
+    # Extend the projected diets to all years, assuming convergence year and demand levels
+    # Original lines 133-138
+    DIET_CONVERGENCE_YEAR <- 9999
+    CONGERENCE_KCALD_CROPS <- 2500
+    CONGERENCE_KCALD_MEAT <- 1000
+
+    L134.pcFood_kcald_R_Dmnd_Y %>%
+      complete(GCAM_region_ID, GCAM_demand,
+               year = unique(c(L134.pcFood_kcald_R_Dmnd_Y$year, HISTORICAL_YEARS, FUTURE_YEARS, DIET_CONVERGENCE_YEAR))) %>%
+      # fill in convergence year (9999) value
+      mutate(value = if_else(year == DIET_CONVERGENCE_YEAR,
+                             if_else(GCAM_demand == "crops", CONGERENCE_KCALD_CROPS, CONGERENCE_KCALD_MEAT),
+                             value)) %>%
+      # interpolate out to that convergence year
+      group_by(GCAM_region_ID, GCAM_demand) %>%
+      mutate(value = approx_fun(year, value)) %>%
+      filter(year %in% c(HISTORICAL_YEARS, FUTURE_YEARS)) ->
+      L134.pcFood_kcald_R_Dmnd_Y
+
+    # Alternative diet scenarios for the SSPs
+    # Original lines 140-
+    # this same logic happens over and over, so *use a function*
+    # scen - scenario
+    create_ssp_diet <- function(L102.pcgdp_thous90USD_Scen_R_Y, scen, demand, L134.pcFood_kcald_R_Dmnd_Y, A_FoodDemand_SSPs) {
+      if(demand == "crops") {
+        a <- 4545    # a and b are parameters for demand scaling function below
+        b <- -0.099
+      } else if(demand == "meat") {
+        a <- 818
+        b <- -2.31
+      } else {
+        stop("Unknown demand ", demand)
+      }
+      L102.pcgdp_thous90USD_Scen_R_Y %>%
+        filter(scenario == scen) %>%
+        mutate(GCAM_demand = demand,
+               value = a / (1 + exp(b * log(value)))) %>%
+        select(GCAM_region_ID, GCAM_demand, year, value) %>%
+        filter(year %in% c(max(HISTORICAL_YEARS), FUTURE_YEARS)) ->
+        ssp_gdp
+
+      ssp_gdp %>%
+        # compute gdp ratios for future years
+        arrange(GCAM_region_ID, GCAM_demand, desc(year)) %>%
+        group_by(GCAM_region_ID, GCAM_demand) %>%
+        mutate(ratio = value / lead(value)) %>%
+        select(-value) %>%
+        # ...and use those ratios to scale future food demand
+        left_join(filter(L134.pcFood_kcald_R_Dmnd_Y, GCAM_demand == demand), ., by = c("GCAM_region_ID", "GCAM_demand", "year")) %>%
+        mutate(ratio = if_else(is.na(ratio), 1.0, ratio)) %>%
+        arrange(GCAM_region_ID, GCAM_demand, year) %>%
+        group_by(GCAM_region_ID, GCAM_demand) %>%
+        # computing the cumulative product, and then multiplying by the last historical year value, is the
+        # same mathematically as multiplying each year's ratio by the previous year's value
+        mutate(ratio = cumprod(ratio),
+               value = if_else(year > max(HISTORICAL_YEARS), ratio * nth(value, which(year == max(HISTORICAL_YEARS))), value)) %>%
```

Comment:
I recognize that I wrote the original code so this will likely be kicked back to me, but I don't understand this line of code (as compared to the for loop below). The for loop below ensures that the year on year growth in food demand doesn't exceed some externally defined ratio. What is this doing?